Allow all directives in line comments

This is an issue to track the development of https://github.com/eslint/rfcs/pull/34.

* [RFC Document](https://github.com/eslint/rfcs/tree/main/designs/2019-allow-all-directives-in-line-comments)

I made a simple implementation, and found ~41 inline comments which were not intended as a directive, but parse as directives.

```bash
➜  eslint git:(issue14575) npm run lint

> eslint@7.27.0 lint
> node Makefile.js lint

Validating JavaScript files

/Users/weiran/repo/github/eslint/lib/rules/newline-before-return.js
  137:51  error  'return' is defined but never used     no-unused-vars
  137:58  error  'at' is defined but never used         no-unused-vars
  137:61  error  'beginning' is defined but never used  no-unused-vars
  137:71  error  'of' is defined but never used         no-unused-vars
  137:74  error  'script' is defined but never used     no-unused-vars

/Users/weiran/repo/github/eslint/tests/lib/rules/id-blacklist.js
  1013:20  error  'declared' is defined but never used       no-unused-vars
  1013:29  error  'in' is defined but never used             no-unused-vars
  1013:32  error  'the' is defined but never used            no-unused-vars
  1013:36  error  'given' is defined but never used          no-unused-vars
  1013:42  error  'source' is defined but never used         no-unused-vars
  1013:49  error  'code' is defined but never used           no-unused-vars
  1013:54  error  'are' is defined but never used            no-unused-vars
  1013:58  error  'not' is defined but never used            no-unused-vars
  1013:62  error  'excluded' is defined but never used       no-unused-vars
  1013:71  error  'from' is defined but never used           no-unused-vars
  1013:76  error  'consideration' is defined but never used  no-unused-vars

/Users/weiran/repo/github/eslint/tests/lib/rules/id-denylist.js
  1013:20  error  'declared' is defined but never used       no-unused-vars
  1013:29  error  'in' is defined but never used             no-unused-vars
  1013:32  error  'the' is defined but never used            no-unused-vars
  1013:36  error  'given' is defined but never used          no-unused-vars
  1013:42  error  'source' is defined but never used         no-unused-vars
  1013:49  error  'code' is defined but never used           no-unused-vars
  1013:54  error  'are' is defined but never used            no-unused-vars
  1013:58  error  'not' is defined but never used            no-unused-vars
  1013:62  error  'excluded' is defined but never used       no-unused-vars
  1013:71  error  'from' is defined but never used           no-unused-vars
  1013:76  error  'consideration' is defined but never used  no-unused-vars

/Users/weiran/repo/github/eslint/tests/lib/rules/no-promise-executor-return.js
   89:19  error  'Promise' is already defined as a built-in global variable  no-redeclare
   89:19  error  'Promise' is defined but never used                         no-unused-vars
   89:27  error  'doesn't' is defined but never used                         no-unused-vars
   89:35  error  'exist' is defined but never used                           no-unused-vars
  100:19  error  'Promise' is already defined as a built-in global variable  no-redeclare
  100:27  error  'is' is defined but never used                              no-unused-vars
  100:30  error  'shadowed' is defined but never used                        no-unused-vars

/Users/weiran/repo/github/eslint/tests/lib/rules/no-setter-return.js
  227:19  error  'object' is defined but never used    no-unused-vars
  227:26  error  'doesn't' is defined but never used   no-unused-vars
  227:34  error  'exist' is defined but never used     no-unused-vars
  235:19  error  'object' is already defined           no-redeclare
  235:26  error  'is' is defined but never used        no-unused-vars
  235:29  error  'shadowed' is defined but never used  no-unused-vars

/Users/weiran/repo/github/eslint/tests/lib/rules/require-await.js
  46:19  error  'await' is defined but never used  no-unused-vars

✖ 41 problems (41 errors, 0 warnings)
```

--- a/docs/rules/no-unused-vars.md
+++ b/docs/rules/no-unused-vars.md
@@ -82,20 +82,24 @@ function getY([, y]) {
 
 ### exported
 
-In environments outside of CommonJS or ECMAScript modules, you may use `var` to create a global variable that may be used by other scripts. You can use the `/* exported variableName */` comment block to indicate that this variable is being exported and therefore should not be considered unused.
+In environments outside of CommonJS or ECMAScript modules, you may use `var` to create a global variable that may be used by other scripts. You can use the comment `/* exported variableName */` or `// exported variableName` to indicate that this variable is being exported and therefore should not be considered unused.
 
-Note that `/* exported */` has no effect for any of the following:
+Note that the comment has no effect for any of the following:
 
 * when the environment is `node` or `commonjs`
 * when `parserOptions.sourceType` is `module`
 * when `ecmaFeatures.globalReturn` is `true`
 
-The line comment `// exported variableName` will not work as `exported` is not line-specific.
-
-Examples of **correct** code for `/* exported variableName */` operation:
+Examples of **correct** code for `exported` operation:
 
 ```js
 /* exported global_var */
+
+var global_var = 42;
+```
+
+```js
+// exported global_var
 
 var global_var = 42;
 ```

--- a/docs/user-guide/configuring/language-options.md
+++ b/docs/user-guide/configuring/language-options.md
@@ -44,10 +44,14 @@ Environments can be specified inside of a file, in configuration files or using 
 
 ### Using configuration comments
 
-To specify environments using a comment inside of your JavaScript file, use the following format:
+To specify environments using a line or block comment inside of your JavaScript file, use the following format:
 
 ```js
 /* eslint-env node, mocha */
+```
+
+```js
+// eslint-env node, mocha
 ```
 
 This enables Node.js and Mocha environments.
@@ -123,10 +127,14 @@ Some of ESLint's core rules rely on knowledge of the global variables available 
 
 ### Using configuration comments
 
-To specify globals using a comment inside of your JavaScript file, use the following format:
+To specify globals using a line or block comment inside of your JavaScript file, use the following format:
 
 ```js
 /* global var1, var2 */
+```
+
+```js
+// global var1, var2
 ```
 
 This defines two global variables, `var1` and `var2`. If you want to optionally specify that these global variables can be written to (rather than only being read), then you can set each with a `"writable"` flag:

--- a/docs/user-guide/configuring/rules.md
+++ b/docs/user-guide/configuring/rules.md
@@ -14,10 +14,14 @@ ESLint comes with a large number of built-in rules and you can add more rules th
 
 ### Using configuration comments
 
-To configure rules inside of a file using configuration comments, use a comment in the following format:
+To configure rules inside of a file using configuration comments, use a line or block comment in the following format:
 
 ```js
 /* eslint eqeqeq: "off", curly: "error" */
+```
+
+```js
+// eslint eqeqeq: "off", curly: "error"
 ```
 
 In this example, [`eqeqeq`](https://eslint.org/docs/rules/eqeqeq) is turned off and [`curly`](.https://eslint.org/docs/rules/curly) is turned on as an error. You can also use the numeric equivalent for the rule severity:
@@ -124,7 +128,7 @@ In these configuration files, the rule `plugin1/rule1` comes from the plugin nam
 
 ### Using configuration comments
 
-To temporarily disable rule warnings in your file, use block comments in the following format:
+To temporarily disable rule warnings in your file, use line or block comments in the following format:
 
 ```js
 /* eslint-disable */
@@ -132,6 +136,14 @@ To temporarily disable rule warnings in your file, use block comments in the fol
 alert('foo');
 
 /* eslint-enable */
+```
+
+```js
+// eslint-disable
+
+alert('foo');
+
+// eslint-enable
 ```
 
 You can also disable or enable warnings for specific rules:
@@ -145,7 +157,7 @@ console.log('bar');
 /* eslint-enable no-alert, no-console */
 ```
 
-To disable rule warnings in an entire file, put a `/* eslint-disable */` block comment at the top of the file:
+To disable rule warnings in an entire file, put a `/* eslint-disable */` or `// eslint-disable` at the top of the file:
 
 ```js
 /* eslint-disable */

--- a/lib/linter/linter.js
+++ b/lib/linter/linter.js
@@ -307,7 +307,7 @@ function getDirectiveComments(filename, ast, ruleMapper, warnInlineConfig) {
             return;
         }
         const directiveText = match[1];
-        const lineCommentSupported = /^eslint-disable-(next-)?line$/u.test(directiveText);
+        const mustBeOnSingleLine = /^eslint-disable-(next-)?line$/u.test(directiveText);
 
         if (warnInlineConfig) {
             const kind = comment.type === "Block" ? `/*${directiveText}*/` : `//${directiveText}`;
@@ -321,7 +321,7 @@ function getDirectiveComments(filename, ast, ruleMapper, warnInlineConfig) {
             return;
         }
 
-        if (lineCommentSupported && comment.loc.start.line !== comment.loc.end.line) {
+        if (mustBeOnSingleLine && comment.loc.start.line !== comment.loc.end.line) {
             const message = `${directiveText} comment should not span multiple lines.`;
 
             problems.push(createLintingProblem({

--- a/lib/linter/linter.js
+++ b/lib/linter/linter.js
@@ -450,7 +450,7 @@ function normalizeEcmaVersion(parser, ecmaVersion) {
 
 const eslintEnvPatterns = [
     /\/\*\s*eslint-env\s(.+?)\*\//gsu, // block comments
-    /\/\/\s*eslint-env\s(.+)/gu // line comments
+    /\/\/[^\S\r\n\u2028\u2029]*eslint-env[^\S\r\n\u2028\u2029](.+)/gu // line comments
 ];
 
 /**

--- a/lib/linter/linter.js
+++ b/lib/linter/linter.js
@@ -309,10 +309,6 @@ function getDirectiveComments(filename, ast, ruleMapper, warnInlineConfig) {
         const directiveText = match[1];
         const lineCommentSupported = /^eslint-disable-(next-)?line$/u.test(directiveText);
 
-        if (comment.type === "Line" && !lineCommentSupported) {
-            return;
-        }
-
         if (warnInlineConfig) {
             const kind = comment.type === "Block" ? `/*${directiveText}*/` : `//${directiveText}`;
 
@@ -452,7 +448,10 @@ function normalizeEcmaVersion(parser, ecmaVersion) {
     return ecmaVersion >= 2015 ? ecmaVersion - 2009 : ecmaVersion;
 }
 
-const eslintEnvPattern = /\/\*\s*eslint-env\s(.+?)\*\//gsu;
+const eslintEnvPatterns = [
+    /\/\*\s*eslint-env\s(.+?)\*\//gsu, // block comments
+    /\/\/\s*eslint-env\s(.+)/gu // line comments
+];
 
 /**
  * Checks whether or not there is a comment which has "eslint-env *" in a given text.
@@ -462,15 +461,16 @@ const eslintEnvPattern = /\/\*\s*eslint-env\s(.+?)\*\//gsu;
 function findEslintEnv(text) {
     let match, retv;
 
-    eslintEnvPattern.lastIndex = 0;
+    for (const eslintEnvPattern of eslintEnvPatterns) {
+        eslintEnvPattern.lastIndex = 0;
 
-    while ((match = eslintEnvPattern.exec(text)) !== null) {
-        retv = Object.assign(
-            retv || {},
-            commentParser.parseListConfig(stripDirectiveComment(match[1]))
-        );
+        while ((match = eslintEnvPattern.exec(text)) !== null) {
+            retv = Object.assign(
+                retv || {},
+                commentParser.parseListConfig(stripDirectiveComment(match[1]))
+            );
+        }
     }
-
     return retv;
 }
 

--- a/lib/rules/newline-before-return.js
+++ b/lib/rules/newline-before-return.js
@@ -134,7 +134,7 @@ module.exports = {
             if (tokenBefore) {
                 lineNumTokenBefore = tokenBefore.loc.end.line;
             } else {
-                lineNumTokenBefore = 0; // global return at beginning of script
+                lineNumTokenBefore = 0; // Global return at beginning of script
             }
 
             return lineNumTokenBefore;

--- a/lib/rules/utils/ast-utils.js
+++ b/lib/rules/utils/ast-utils.js
@@ -853,7 +853,7 @@ module.exports = {
     isDirectiveComment(node) {
         const comment = node.value.trim();
 
-        return /^(eslint[- ]|(globals?|exported) )/u.test(comment);
+        return /^(?:eslint[- ]|(?:globals?|exported) )/u.test(comment);
     },
 
     /**

--- a/lib/rules/utils/ast-utils.js
+++ b/lib/rules/utils/ast-utils.js
@@ -853,14 +853,7 @@ module.exports = {
     isDirectiveComment(node) {
         const comment = node.value.trim();
 
-        return (
-            node.type === "Line" && comment.indexOf("eslint-") === 0 ||
-            node.type === "Block" && (
-                comment.indexOf("global ") === 0 ||
-                comment.indexOf("eslint ") === 0 ||
-                comment.indexOf("eslint-") === 0
-            )
-        );
+        return /^(eslint[- ])|(global )/u.test(comment);
     },
 
     /**

--- a/lib/rules/utils/ast-utils.js
+++ b/lib/rules/utils/ast-utils.js
@@ -853,7 +853,7 @@ module.exports = {
     isDirectiveComment(node) {
         const comment = node.value.trim();
 
-        return /^(eslint[- ]|global )/u.test(comment);
+        return /^(eslint[- ]|(globals?|exported) )/u.test(comment);
     },
 
     /**

--- a/lib/rules/utils/ast-utils.js
+++ b/lib/rules/utils/ast-utils.js
@@ -853,7 +853,7 @@ module.exports = {
     isDirectiveComment(node) {
         const comment = node.value.trim();
 
-        return /^(eslint[- ])|(global )/u.test(comment);
+        return /^(eslint[- ]|global )/u.test(comment);
     },
 
     /**

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -5570,12 +5570,50 @@ var a = "test2";
             assert.strictEqual(messages.length, 0);
         });
 
-        it("//eslint-env", () => {
-            const code = `//${ESLINT_ENV} browser\nwindow;`;
-            const config = { rules: { "no-undef": 2 } };
-            const messages = linter.verify(code, config, filename);
+        describe("//eslint-env", () => {
+            const linebreaks = ["\n", "\r\n", "\r", "\u2028", "\u2029"];
 
-            assert.strictEqual(messages.length, 0);
+            it("enable one env with different line breaks", () => {
+                const codes = [
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser${linebreak}window;`),
+                    ...linebreaks.map(linebreak => `//  ${ESLINT_ENV}  browser${linebreak}window;`)
+                ];
+
+                for (const code of codes) {
+                    const config = { rules: { "no-undef": 2 } };
+                    const messages = linter.verify(code, config, filename);
+
+                    assert.strictEqual(messages.length, 0);
+                }
+            });
+
+            it("multiple envs enabled with different line breaks", () => {
+                const codes = linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`);
+
+                for (const code of codes) {
+                    const config = { rules: { "no-undef": 2 } };
+                    const messages = linter.verify(code, config, filename);
+
+                    assert.strictEqual(messages.length, 0);
+                }
+            });
+
+
+            it("no env enabled with different linebreaks", () => {
+                const codes = [
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV}${linebreak}browser${linebreak}window;`),
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV}browser${linebreak}window;window;`)
+                ];
+
+                for (const code of codes) {
+                    const config = { rules: { "no-undef": 2 } };
+                    const messages = linter.verify(code, config, filename);
+
+                    assert.strictEqual(messages.length, 2);
+                }
+            });
+
+
         });
     });
 });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -23,6 +23,7 @@ const { Linter } = require("../../../lib/linter");
 
 const TEST_CODE = "var answer = 6 * 7;",
     BROKEN_TEST_CODE = "var;";
+const linebreaks = ["\n", "\r\n", "\r", "\u2028", "\u2029"];
 
 //------------------------------------------------------------------------------
 // Helpers
@@ -5525,28 +5526,38 @@ var a = "test2";
 
     // https://github.com/eslint/eslint/issues/14575
     describe("directives in line comments", () => {
+        let messages;
+
         it("//eslint-disable", () => {
-            let code = "//eslint-disable no-debugger\ndebugger;";
-            let messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+            const codes = linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}debugger;`);
 
-            assert.strictEqual(messages.length, 0);
+            for (const code of codes) {
+                messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+                assert.strictEqual(messages.length, 0);
 
-            code = "//eslint-disable no-debugger\ndebugger;";
-            messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+                messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+                assert.strictEqual(messages.length, 0);
+            }
+        });
 
-            assert.strictEqual(messages.length, 0);
+        it("//eslint-enable", () => {
+            const codes = linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}//eslint-enable no-debugger${linebreak}debugger;`);
+
+            for (const code of codes) {
+                messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+                assert.strictEqual(messages.length, 1);
+            }
         });
 
         it("//eslint", () => {
-            let code = "//eslint no-debugger:'error'\ndebugger;";
-            let messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+            const codes = linebreaks.map(linebreak => `//eslint no-debugger:'error'${linebreak}debugger;`);
 
-            assert.strictEqual(messages.length, 1);
-
-            code = "//eslint no-debugger:'error'\ndebugger;";
-            messages = linter.verify(code, { rules: { "no-debugger": 2 } });
-
-            assert.strictEqual(messages.length, 1);
+            for (const code of codes) {
+                messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+                assert.strictEqual(messages.length, 1);
+                messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+                assert.strictEqual(messages.length, 1);
+            }
         });
 
         it("//global(s)", () => {
@@ -5557,21 +5568,19 @@ var a = "test2";
             ];
 
             for (const code of codes) {
-                const messages = linter.verify(code, config, filename);
-
+                messages = linter.verify(code, config, filename);
                 assert.strictEqual(messages.length, 0);
             }
         });
         it("//exported", () => {
             const code = "//exported foo\nvar foo = 0;";
             const config = { rules: { "no-unused-vars": 2 } };
-            const messages = linter.verify(code, config, filename);
 
+            messages = linter.verify(code, config, filename);
             assert.strictEqual(messages.length, 0);
         });
 
         describe("//eslint-env", () => {
-            const linebreaks = ["\n", "\r\n", "\r", "\u2028", "\u2029"];
             const config = { rules: { "no-undef": 2 } };
 
             it("enable one env with different line breaks", () => {
@@ -5581,8 +5590,7 @@ var a = "test2";
                 ];
 
                 for (const code of codes) {
-                    const messages = linter.verify(code, config, filename);
-
+                    messages = linter.verify(code, config, filename);
                     assert.strictEqual(messages.length, 0);
                 }
             });
@@ -5591,8 +5599,7 @@ var a = "test2";
                 const codes = linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`);
 
                 for (const code of codes) {
-                    const messages = linter.verify(code, config, filename);
-
+                    messages = linter.verify(code, config, filename);
                     assert.strictEqual(messages.length, 0);
                 }
             });
@@ -5604,8 +5611,7 @@ var a = "test2";
                 ];
 
                 for (const code of codes) {
-                    const messages = linter.verify(code, config, filename);
-
+                    messages = linter.verify(code, config, filename);
                     assert.strictEqual(messages.length, 2);
                 }
             });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -5572,6 +5572,7 @@ var a = "test2";
 
         describe("//eslint-env", () => {
             const linebreaks = ["\n", "\r\n", "\r", "\u2028", "\u2029"];
+            const config = { rules: { "no-undef": 2 } };
 
             it("enable one env with different line breaks", () => {
                 const codes = [
@@ -5580,7 +5581,6 @@ var a = "test2";
                 ];
 
                 for (const code of codes) {
-                    const config = { rules: { "no-undef": 2 } };
                     const messages = linter.verify(code, config, filename);
 
                     assert.strictEqual(messages.length, 0);
@@ -5591,7 +5591,6 @@ var a = "test2";
                 const codes = linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`);
 
                 for (const code of codes) {
-                    const config = { rules: { "no-undef": 2 } };
                     const messages = linter.verify(code, config, filename);
 
                     assert.strictEqual(messages.length, 0);
@@ -5605,7 +5604,6 @@ var a = "test2";
                 ];
 
                 for (const code of codes) {
-                    const config = { rules: { "no-undef": 2 } };
                     const messages = linter.verify(code, config, filename);
 
                     assert.strictEqual(messages.length, 2);

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -1360,15 +1360,16 @@ describe("Linter", () => {
     describe("when evaluating code containing a line comment", () => {
         const code = "//global a \n function f() {}";
 
-        it("should not introduce a global variable", () => {
+        it("should introduce a global variable", () => {
             const config = { rules: { checker: "error" } };
             let spy;
 
             linter.defineRule("checker", context => {
                 spy = sinon.spy(() => {
                     const scope = context.getScope();
+                    const a = getVariable(scope, "a");
 
-                    assert.strictEqual(getVariable(scope, "a"), null);
+                    assert.strictEqual(a.eslintExplicitGlobal, true);
                 });
 
                 return { Program: spy };
@@ -5517,6 +5518,62 @@ var a = "test2";
         it("should not pass any default parserOptions to the parser", () => {
             linter.defineParser("throws-with-options", testParsers.throwsWithOptions);
             const messages = linter.verify(";", { parser: "throws-with-options" }, "filename");
+
+            assert.strictEqual(messages.length, 0);
+        });
+    });
+
+    // https://github.com/eslint/eslint/issues/14575
+    describe("directives in line comments", () => {
+        it("//eslint-disable", () => {
+            let code = "//eslint-disable no-debugger\ndebugger;";
+            let messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+
+            assert.strictEqual(messages.length, 0);
+
+            code = "//eslint-disable no-debugger\ndebugger;";
+            messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+
+            assert.strictEqual(messages.length, 0);
+        });
+
+        it("//eslint", () => {
+            let code = "//eslint no-debugger:'error'\ndebugger;";
+            let messages = linter.verify(code, { rules: { "no-debugger": 0 } });
+
+            assert.strictEqual(messages.length, 1);
+
+            code = "//eslint no-debugger:'error'\ndebugger;";
+            messages = linter.verify(code, { rules: { "no-debugger": 2 } });
+
+            assert.strictEqual(messages.length, 1);
+        });
+
+        it("//global(s)", () => {
+            let code = "//globals foo: true\nfoo;";
+            let config = { rules: { "no-undef": 2 } };
+            let messages = linter.verify(code, config, filename);
+
+            assert.strictEqual(messages.length, 0);
+
+            code = "//global foo: true\nfoo;";
+            config = { rules: { "no-undef": 2 } };
+            messages = linter.verify(code, config, filename);
+
+            assert.strictEqual(messages.length, 0);
+        });
+        it("//exported", () => {
+            const code = "//exported foo\nvar foo = 0;";
+            const config = { rules: { "no-unused-vars": 2 } };
+            const messages = linter.verify(code, config, filename);
+
+            assert.strictEqual(messages.length, 0);
+        });
+
+        it("//eslint-env", () => {
+            const code = `//${ESLINT_ENV} browser\nwindow;`;
+            const config = { rules: { "no-undef": 2 } };
+            const messages = linter.verify(code, config, filename);
 
             assert.strictEqual(messages.length, 0);
         });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -5550,17 +5550,17 @@ var a = "test2";
         });
 
         it("//global(s)", () => {
-            let code = "//globals foo: true\nfoo;";
-            let config = { rules: { "no-undef": 2 } };
-            let messages = linter.verify(code, config, filename);
+            const config = { rules: { "no-undef": 2 } };
+            const codes = [
+                "//globals foo: true\nfoo;",
+                "//global foo: true\nfoo;"
+            ];
 
-            assert.strictEqual(messages.length, 0);
+            for (const code of codes) {
+                const messages = linter.verify(code, config, filename);
 
-            code = "//global foo: true\nfoo;";
-            config = { rules: { "no-undef": 2 } };
-            messages = linter.verify(code, config, filename);
-
-            assert.strictEqual(messages.length, 0);
+                assert.strictEqual(messages.length, 0);
+            }
         });
         it("//exported", () => {
             const code = "//exported foo\nvar foo = 0;";
@@ -5598,7 +5598,6 @@ var a = "test2";
                 }
             });
 
-
             it("no env enabled with different linebreaks", () => {
                 const codes = [
                     ...linebreaks.map(linebreak => `//${ESLINT_ENV}${linebreak}browser${linebreak}window;`),
@@ -5612,8 +5611,6 @@ var a = "test2";
                     assert.strictEqual(messages.length, 2);
                 }
             });
-
-
         });
     });
 });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -5586,7 +5586,8 @@ var a = "test2";
             it("enable one env with different line breaks", () => {
                 const codes = [
                     ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser${linebreak}window;`),
-                    ...linebreaks.map(linebreak => `//  ${ESLINT_ENV}  browser${linebreak}window;`)
+                    ...linebreaks.map(linebreak => `//  ${ESLINT_ENV}  browser${linebreak}window;`),
+                    `window;//${ESLINT_ENV} browser` // no linebreaks
                 ];
 
                 for (const code of codes) {
@@ -5596,7 +5597,10 @@ var a = "test2";
             });
 
             it("multiple envs enabled with different line breaks", () => {
-                const codes = linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`);
+                const codes = [
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`),
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser${linebreak}//${ESLINT_ENV} es6${linebreak}window;Promise;`)
+                ];
 
                 for (const code of codes) {
                     messages = linter.verify(code, config, filename);

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -5529,7 +5529,10 @@ var a = "test2";
         let messages;
 
         it("//eslint-disable", () => {
-            const codes = linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}debugger;`);
+            const codes = [
+                ...linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}debugger;`),
+                ...linebreaks.map(linebreak => `// eslint-disable no-debugger${linebreak}debugger;`)
+            ];
 
             for (const code of codes) {
                 messages = linter.verify(code, { rules: { "no-debugger": 2 } });
@@ -5541,7 +5544,10 @@ var a = "test2";
         });
 
         it("//eslint-enable", () => {
-            const codes = linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}//eslint-enable no-debugger${linebreak}debugger;`);
+            const codes = [
+                ...linebreaks.map(linebreak => `//eslint-disable no-debugger${linebreak}//eslint-enable no-debugger${linebreak}debugger;`),
+                ...linebreaks.map(linebreak => `// eslint-disable no-debugger${linebreak}//eslint-enable no-debugger${linebreak}debugger;`)
+            ];
 
             for (const code of codes) {
                 messages = linter.verify(code, { rules: { "no-debugger": 2 } });
@@ -5550,7 +5556,10 @@ var a = "test2";
         });
 
         it("//eslint", () => {
-            const codes = linebreaks.map(linebreak => `//eslint no-debugger:'error'${linebreak}debugger;`);
+            const codes = [
+                ...linebreaks.map(linebreak => `//eslint no-debugger:'error'${linebreak}debugger;`),
+                ...linebreaks.map(linebreak => `// eslint no-debugger:'error'${linebreak}debugger;`)
+            ];
 
             for (const code of codes) {
                 messages = linter.verify(code, { rules: { "no-debugger": 0 } });
@@ -5563,8 +5572,10 @@ var a = "test2";
         it("//global(s)", () => {
             const config = { rules: { "no-undef": 2 } };
             const codes = [
-                "//globals foo: true\nfoo;",
-                "//global foo: true\nfoo;"
+                ...linebreaks.map(linebreak => `//globals foo: true${linebreak}foo;`),
+                ...linebreaks.map(linebreak => `// globals foo: true${linebreak}foo;`),
+                ...linebreaks.map(linebreak => `//global foo: true${linebreak}foo;`),
+                ...linebreaks.map(linebreak => `// global foo: true${linebreak}foo;`)
             ];
 
             for (const code of codes) {
@@ -5573,11 +5584,16 @@ var a = "test2";
             }
         });
         it("//exported", () => {
-            const code = "//exported foo\nvar foo = 0;";
+            const codes = [
+                ...linebreaks.map(linebreak => `//exported foo${linebreak}var foo = 0;`),
+                ...linebreaks.map(linebreak => `// exported foo${linebreak}var foo = 0;`)
+            ];
             const config = { rules: { "no-unused-vars": 2 } };
 
-            messages = linter.verify(code, config, filename);
-            assert.strictEqual(messages.length, 0);
+            for (const code of codes) {
+                messages = linter.verify(code, config, filename);
+                assert.strictEqual(messages.length, 0);
+            }
         });
 
         describe("//eslint-env", () => {
@@ -5599,7 +5615,9 @@ var a = "test2";
             it("multiple envs enabled with different line breaks", () => {
                 const codes = [
                     ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser,es6${linebreak}window;Promise;`),
-                    ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser${linebreak}//${ESLINT_ENV} es6${linebreak}window;Promise;`)
+                    ...linebreaks.map(linebreak => `// ${ESLINT_ENV} browser,es6${linebreak}window;Promise;`),
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV} browser${linebreak}//${ESLINT_ENV} es6${linebreak}window;Promise;`),
+                    ...linebreaks.map(linebreak => `// ${ESLINT_ENV} browser${linebreak}//${ESLINT_ENV} es6${linebreak}window;Promise;`)
                 ];
 
                 for (const code of codes) {
@@ -5611,7 +5629,9 @@ var a = "test2";
             it("no env enabled with different linebreaks", () => {
                 const codes = [
                     ...linebreaks.map(linebreak => `//${ESLINT_ENV}${linebreak}browser${linebreak}window;`),
-                    ...linebreaks.map(linebreak => `//${ESLINT_ENV}browser${linebreak}window;window;`)
+                    ...linebreaks.map(linebreak => `// ${ESLINT_ENV}${linebreak}browser${linebreak}window;`),
+                    ...linebreaks.map(linebreak => `//${ESLINT_ENV}browser${linebreak}window;window;`),
+                    ...linebreaks.map(linebreak => `// ${ESLINT_ENV}browser${linebreak}window;window;`)
                 ];
 
                 for (const code of codes) {

--- a/tests/lib/rules/id-blacklist.js
+++ b/tests/lib/rules/id-blacklist.js
@@ -1010,7 +1010,7 @@ ruleTester.run("id-blacklist", rule, {
             ]
         },
 
-        // globals declared in the given source code are not excluded from consideration
+        // Globals declared in the given source code are not excluded from consideration
         {
             code: "const foo = 1; bar = foo;",
             options: ["foo"],

--- a/tests/lib/rules/id-denylist.js
+++ b/tests/lib/rules/id-denylist.js
@@ -1010,7 +1010,7 @@ ruleTester.run("id-denylist", rule, {
             ]
         },
 
-        // globals declared in the given source code are not excluded from consideration
+        // Globals declared in the given source code are not excluded from consideration
         {
             code: "const foo = 1; bar = foo;",
             options: ["foo"],

--- a/tests/lib/rules/no-inline-comments.js
+++ b/tests/lib/rules/no-inline-comments.js
@@ -40,6 +40,12 @@ ruleTester.run("no-inline-comments", rule, {
         "var a = 1; // eslint-disable-line no-debugger",
         "var a = 1; /* eslint-disable-line no-debugger */",
 
+        // https://github.com/eslint/eslint/issues/14575
+        "var i;//eslint no-debugger:2",
+        "var i;//eslint-disable no-debugger",
+        "var i;//eslint-enable no-debugger",
+        "var i;//global foo",
+
         // JSX exception
         `var a = (
             <div>

--- a/tests/lib/rules/no-inline-comments.js
+++ b/tests/lib/rules/no-inline-comments.js
@@ -42,9 +42,13 @@ ruleTester.run("no-inline-comments", rule, {
 
         // https://github.com/eslint/eslint/issues/14575
         "var i;//eslint no-debugger:2",
+        "var i;// eslint no-debugger:2",
         "var i;//eslint-disable no-debugger",
+        "var i;// eslint-disable no-debugger",
         "var i;//eslint-enable no-debugger",
+        "var i;// eslint-enable no-debugger",
         "var i;//global foo",
+        "var i;// global foo",
 
         // JSX exception
         `var a = (

--- a/tests/lib/rules/no-promise-executor-return.js
+++ b/tests/lib/rules/no-promise-executor-return.js
@@ -86,7 +86,7 @@ ruleTester.run("no-promise-executor-return", rule, {
         "new Promise(foo, (resolve, reject) => { return 1; });",
         "new Promise(foo, (resolve, reject) => 1);",
 
-        // global Promise doesn't exist
+        // Global Promise doesn't exist
         "/* globals Promise:off */ new Promise(function (resolve, reject) { return 1; });",
         {
             code: "new Promise((resolve, reject) => { return 1; });",
@@ -97,7 +97,7 @@ ruleTester.run("no-promise-executor-return", rule, {
             env: { es6: false }
         },
 
-        // global Promise is shadowed
+        // Global Promise is shadowed
         "let Promise; new Promise(function (resolve, reject) { return 1; });",
         "function f() { new Promise((resolve, reject) => { return 1; }); var Promise; }",
         "function f(Promise) { new Promise((resolve, reject) => 1); }",

--- a/tests/lib/rules/no-setter-return.js
+++ b/tests/lib/rules/no-setter-return.js
@@ -224,7 +224,7 @@ ruleTester.run("no-setter-return", rule, {
         },
         "object.create(foo, { bar: { set: function(val) { return 1; } } })",
 
-        // global object doesn't exist
+        // Global object doesn't exist
         "Reflect.defineProperty(foo, 'bar', { set(val) { if (val) { return 1; } } })",
         "/* globals Object:off */ Object.defineProperty(foo, 'bar', { set(val) { return 1; } })",
         {
@@ -232,7 +232,7 @@ ruleTester.run("no-setter-return", rule, {
             globals: { Object: "off" }
         },
 
-        // global object is shadowed
+        // Global object is shadowed
         "let Object; Object.defineProperty(foo, 'bar', { set(val) { return 1; } })",
         {
             code: "function f() { Reflect.defineProperty(foo, 'bar', { set(val) { if (val) { return 1; } } }); var Reflect;}",

--- a/tests/lib/rules/no-warning-comments.js
+++ b/tests/lib/rules/no-warning-comments.js
@@ -37,7 +37,13 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "// special regex characters don't cause problems", options: [{ terms: ["[aeiou]"], location: "anywhere" }] },
         "/*eslint no-warning-comments: [2, { \"terms\": [\"todo\", \"fixme\", \"any other term\"], \"location\": \"anywhere\" }]*/\n\nvar x = 10;\n",
         { code: "/*eslint no-warning-comments: [2, { \"terms\": [\"todo\", \"fixme\", \"any other term\"], \"location\": \"anywhere\" }]*/\n\nvar x = 10;\n", options: [{ location: "anywhere" }] },
-        { code: "foo", options: [{ terms: ["foo-bar"] }] }
+        { code: "foo", options: [{ terms: ["foo-bar"] }] },
+
+        // https://github.com/eslint/eslint/issues/14575
+        "//eslint-enable no-debugger",
+        "//eslint-disable no-debugger",
+        "//global foo",
+        "//eslint no-debugger: 0"
     ],
     invalid: [
         {

--- a/tests/lib/rules/no-warning-comments.js
+++ b/tests/lib/rules/no-warning-comments.js
@@ -40,10 +40,7 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "foo", options: [{ terms: ["foo-bar"] }] },
 
         // https://github.com/eslint/eslint/issues/14575
-        "//eslint-enable no-debugger",
-        "//eslint-disable no-debugger",
-        "//global foo",
-        "//eslint no-debugger: 0"
+        "//eslint no-warning-comments: [2, {terms: [\"eslint\"]}]"
     ],
     invalid: [
         {

--- a/tests/lib/rules/require-await.js
+++ b/tests/lib/rules/require-await.js
@@ -43,7 +43,7 @@ ruleTester.run("require-await", rule, {
         // for-await-of
         "async function foo() { for await (x of xs); }",
 
-        // global await
+        // Global await
         {
             code: "await foo()",
             parser: require.resolve("../../fixtures/parsers/typescript-parsers/global-await")

--- a/tests/lib/rules/utils/ast-utils.js
+++ b/tests/lib/rules/utils/ast-utils.js
@@ -205,8 +205,7 @@ describe("ast-utils", () => {
             const code = [
                 "// lalala I'm a normal comment",
                 "// trying to confuse eslint ",
-                "//trying to confuse eslint-directive-detection",
-                "//eslint is awesome"
+                "//trying to confuse eslint-directive-detection"
             ].join("\n");
             const ast = espree.parse(code, ESPREE_CONFIG);
             const sourceCode = new SourceCode(code, ast);
@@ -234,7 +233,13 @@ describe("ast-utils", () => {
                 "// eslint-disable-line no-undef",
                 "// eslint-secret-directive 4 8 15 16 23 42   ",
                 "// eslint-directive-without-argument",
-                "//eslint-directive-without-padding"
+                "//eslint-directive-without-padding",
+
+                // https://github.com/eslint/eslint/issues/14575
+                "//eslint foo:0",
+                "//global foo",
+                "//eslint-enable foo",
+                "//eslint-disable foo"
             ].join("\n");
             const ast = espree.parse(code, ESPREE_CONFIG);
             const sourceCode = new SourceCode(code, ast);

--- a/tests/lib/rules/utils/ast-utils.js
+++ b/tests/lib/rules/utils/ast-utils.js
@@ -205,7 +205,14 @@ describe("ast-utils", () => {
             const code = [
                 "// lalala I'm a normal comment",
                 "// trying to confuse eslint ",
-                "//trying to confuse eslint-directive-detection"
+                "//trying to confuse eslint-directive-detection",
+                "//xxeslint-enable foo",
+                "//xxeslint-disable foo",
+                "//xxeslint foo: 2",
+                "//xxglobal foo",
+                "//xxglobals foo",
+                "//xxexported foo",
+                ""
             ].join("\n");
             const ast = espree.parse(code, ESPREE_CONFIG);
             const sourceCode = new SourceCode(code, ast);

--- a/tests/lib/rules/utils/ast-utils.js
+++ b/tests/lib/rules/utils/ast-utils.js
@@ -211,7 +211,7 @@ describe("ast-utils", () => {
                 "//xxeslint foo: 2",
                 "//xxglobal foo",
                 "//xxglobals foo",
-                "//xxexported foo",
+                "//xxexported foo"
             ].join("\n");
             const ast = espree.parse(code, ESPREE_CONFIG);
             const sourceCode = new SourceCode(code, ast);

--- a/tests/lib/rules/utils/ast-utils.js
+++ b/tests/lib/rules/utils/ast-utils.js
@@ -212,7 +212,6 @@ describe("ast-utils", () => {
                 "//xxglobal foo",
                 "//xxglobals foo",
                 "//xxexported foo",
-                ""
             ].join("\n");
             const ast = espree.parse(code, ESPREE_CONFIG);
             const sourceCode = new SourceCode(code, ast);

